File: contracts/agency.mjs
export const AGENCY_CONTRACT_VERSION = "actuation.agency/v1";

const DETERMINATION_KINDS = new Set([
  "self-differentiation",
  "delegation",
  "derivation",
  "federation",
]);
const RETURN_MODES = new Set(["required", "optional", "autonomous-termination"]);
const RECOGNITION_STATES = new Set(["pending", "recognised", "rejected"]);
const MUTATION_STATES = new Set(["not-applied", "applied"]);
const METAGENCY_OPERATIONS = new Set([
  "determine-agency",
  "configure-agency",
  "actualise-agency",
  "reintegrate-return",
]);

function record(value, name) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError(`${name} must be an object`);
  }
  return value;
}

function ref(value, name, { optional = false } = {}) {
  if (value == null && optional) return;
  if (typeof value !== "string" || value.trim() === "") {
    throw new TypeError(`${name} must be a non-empty string ref`);
  }
}

function refArray(value, name, { optional = false, nonEmpty = false } = {}) {
  if (value == null && optional) return;
  if (!Array.isArray(value) || value.some((item) => typeof item !== "string" || item.trim() === "")) {
    throw new TypeError(`${name} must be an array of non-empty refs`);
  }
  if (nonEmpty && value.length === 0) throw new TypeError(`${name} must not be empty`);
}

function schema(value, name) {
  if (value !== AGENCY_CONTRACT_VERSION) {
    throw new TypeError(`${name}.schema must equal ${AGENCY_CONTRACT_VERSION}`);
  }
}

export function validateWorldBinding(input) {
  const binding = record(input, "WorldBinding");
  schema(binding.schema, "WorldBinding");
  ref(binding.binding_ref, "WorldBinding.binding_ref");
  ref(binding.agent_ref, "WorldBinding.agent_ref");
  ref(binding.agency_ref, "WorldBinding.agency_ref");
  ref(binding.world_ref, "WorldBinding.world_ref");
  ref(binding.scope_ref, "WorldBinding.scope_ref");
  ref(binding.purpose_ref, "WorldBinding.purpose_ref", { optional: true });
  ref(binding.continuity_ref, "WorldBinding.continuity_ref", { optional: true });
  ref(binding.determining_agency_ref, "WorldBinding.determining_agency_ref", { optional: true });
  ref(binding.return_relation_ref, "WorldBinding.return_relation_ref", { optional: true });
  refArray(binding.bounds_refs, "WorldBinding.bounds_refs", { optional: true });
  refArray(binding.authority_refs, "WorldBinding.authority_refs", { optional: true });

  if (binding.constraints != null) {
    const constraints = record(binding.constraints, "WorldBinding.constraints");
    refArray(constraints.human_authored_refs, "WorldBinding.constraints.human_authored_refs", { optional: true });
    refArray(constraints.security_policy_refs, "WorldBinding.constraints.security_policy_refs", { optional: true });
    refArray(constraints.evidence_refs, "WorldBinding.constraints.evidence_refs", { optional: true });
    refArray(constraints.external_reality_refs, "WorldBinding.constraints.external_reality_refs", { optional: true });
  }
  return binding;
}

export function validateRootScope(input) {
  const scope = record(input, "RootScope");
  schema(scope.schema, "RootScope");
  ref(scope.scope_ref, "RootScope.scope_ref");
  ref(scope.enclosing_world_ref, "RootScope.enclosing_world_ref");
  return scope;
}

/** Root Agency is a positional relation, never an Agent subtype. */
export function isRootAgency(worldBinding, rootScope) {
  const binding = validateWorldBinding(worldBinding);
  const scope = validateRootScope(rootScope);
  return binding.scope_ref === scope.scope_ref && binding.world_ref === scope.enclosing_world_ref;
}

export function validateMetagencyGrant(input) {
  const grant = record(input, "MetagencyGrant");
  schema(grant.schema, "MetagencyGrant");
  ref(grant.grant_ref, "MetagencyGrant.grant_ref");
  ref(grant.agency_ref, "MetagencyGrant.agency_ref");
  ref(grant.world_binding_ref, "MetagencyGrant.world_binding_ref");
  ref(grant.authority_ref, "MetagencyGrant.authority_ref");
  refArray(grant.bounds_refs, "MetagencyGrant.bounds_refs", { optional: true });
  if (!Array.isArray(grant.operations) || grant.operations.length === 0) {
    throw new TypeError("MetagencyGrant.operations must be a non-empty array");
  }
  for (const operation of grant.operations) {
    if (!METAGENCY_OPERATIONS.has(operation)) {
      throw new TypeError(`MetagencyGrant.operations contains unsupported operation: ${operation}`);
    }
  }
  return grant;
}

export function validateDetermination(input) {
  const determination = record(input, "Determination");
  schema(determination.schema, "Determination");
  ref(determination.determination_ref, "Determination.determination_ref");
  if (!DETERMINATION_KINDS.has(determination.kind)) {
    throw new TypeError("Determination.kind must distinguish self-differentiation, delegation, derivation or federation");
  }
  ref(determination.determining_agency_ref, "Determination.determining_agency_ref");
  ref(determination.differentiated_agency_ref, "Determination.differentiated_agency_ref");
  ref(determination.world_binding_ref, "Determination.world_binding_ref");
  ref(determination.parent_determination_ref, "Determination.parent_determination_ref", { optional: true });
  refArray(determination.bounds_refs, "Determination.bounds_refs", { nonEmpty: true });
  refArray(determination.authority_refs, "Determination.authority_refs", { optional: true });

  const autonomy = record(determination.delegated_autonomy, "Determination.delegated_autonomy");
  refArray(autonomy.allowed_action_refs, "Determination.delegated_autonomy.allowed_action_refs", { optional: true });
  refArray(autonomy.denied_action_refs, "Determination.delegated_autonomy.denied_action_refs", { optional: true });
  if (typeof autonomy.may_determine_within_bounds !== "boolean") {
    throw new TypeError("Determination.delegated_autonomy.may_determine_within_bounds must be boolean");
  }

  const returnPolicy = record(determination.return_policy, "Determination.return_policy");
  if (!RETURN_MODES.has(returnPolicy.mode)) {
    throw new TypeError("Determination.return_policy.mode must be required, optional or autonomous-termination");
  }
  ref(returnPolicy.return_relation_ref, "Determination.return_policy.return_relation_ref", {
    optional: returnPolicy.mode === "autonomous-termination",
  });

  if (determination.kind === "federation" && determination.authority_refs?.length) {
    throw new TypeError("Federation cannot silently carry determining authority; use an explicit delegation if authority is granted");
  }
  return determination;
}

export function validateReturn(input) {
  const returned = record(input, "Return");
  schema(returned.schema, "Return");
  ref(returned.return_ref, "Return.return_ref");
  ref(returned.determination_ref, "Return.determination_ref");
  ref(returned.from_agency_ref, "Return.from_agency_ref");
  ref(returned.to_agency_ref, "Return.to_agency_ref");
  refArray(returned.difference_refs, "Return.difference_refs", { nonEmpty: true });
  refArray(returned.artifact_refs, "Return.artifact_refs", { optional: true });
  refArray(returned.claim_refs, "Return.claim_refs", { optional: true });
  refArray(returned.evidence_refs, "Return.evidence_refs", { optional: true });

  const provenance = record(returned.provenance, "Return.provenance");
  refArray(provenance.agency_lineage_refs, "Return.provenance.agency_lineage_refs", { nonEmpty: true });
  refArray(provenance.material_refs, "Return.provenance.material_refs", { optional: true });
  refArray(provenance.external_source_refs, "Return.provenance.external_source_refs", { optional: true });

  if (typeof returned.received !== "boolean") throw new TypeError("Return.received must be boolean");
  if (!RECOGNITION_STATES.has(returned.recognition_state)) {
    throw new TypeError("Return.recognition_state must be pending, recognised or rejected");
  }
  if (!MUTATION_STATES.has(returned.world_mutation_state)) {
    throw new TypeError("Return.world_mutation_state must be not-applied or applied");
  }
  if (!returned.received && returned.recognition_state !== "pending") {
    throw new TypeError("A Return cannot be recognised/rejected before it is received");
  }
  if (returned.world_mutation_state === "applied" && returned.recognition_state !== "recognised") {
    throw new TypeError("World mutation requires an explicitly recognised Return");
  }
  return returned;
}

export function validateDeterminationLineage(determinations) {
  if (!Array.isArray(determinations) || determinations.length === 0) {
    throw new TypeError("Determination lineage must be a non-empty array");
  }
  const validated = determinations.map(validateDetermination);
  const byRef = new Map(validated.map((item) => [item.determination_ref, item]));
  for (const item of validated) {
    if (!item.parent_determination_ref) continue;
    const parent = byRef.get(item.parent_determination_ref);
    if (!parent) throw new TypeError(`Missing parent determination ${item.parent_determination_ref}`);
    if (parent.differentiated_agency_ref !== item.determining_agency_ref) {
      throw new TypeError("Recursive determination lineage must continue through the agency determined by its parent");
    }
    if (!parent.delegated_autonomy.may_determine_within_bounds) {
      throw new TypeError("A differentiated Agency may govern downward only when delegated autonomy permits determination");
    }
  }
  return validated;
}

export function agencyReadModel({ binding, root_scope, metagency_grants = [], determinations = [], returns = [] }) {
  const worldBinding = validateWorldBinding(binding);
  const grants = metagency_grants.map(validateMetagencyGrant).filter((grant) => grant.agency_ref === worldBinding.agency_ref);
  const lineage = determinations.length ? validateDeterminationLineage(determinations) : [];
  const returnModels = returns.map(validateReturn);
  return {
    schema: AGENCY_CONTRACT_VERSION,
    agency_ref: worldBinding.agency_ref,
    agent_ref: worldBinding.agent_ref,
    world_binding_ref: worldBinding.binding_ref,
    world_ref: worldBinding.world_ref,
    scope_ref: worldBinding.scope_ref,
    root_for_scope: root_scope ? isRootAgency(worldBinding, root_scope) : false,
    metagency: {
      available: grants.length > 0,
      operations: [...new Set(grants.flatMap((grant) => grant.operations))].sort(),
      grant_refs: grants.map((grant) => grant.grant_ref),
    },
    governing_determination_refs: lineage
      .filter((item) => item.determining_agency_ref === worldBinding.agency_ref)
      .map((item) => item.determination_ref),
    governed_by_determination_refs: lineage
      .filter((item) => item.differentiated_agency_ref === worldBinding.agency_ref)
      .map((item) => item.determination_ref),
    returns: {
      pending: returnModels.filter((item) => item.to_agency_ref === worldBinding.agency_ref && !item.received).map((item) => item.return_ref),
      received: returnModels.filter((item) => item.to_agency_ref === worldBinding.agency_ref && item.received).map((item) => item.return_ref),
      recognised: returnModels.filter((item) => item.to_agency_ref === worldBinding.agency_ref && item.recognition_state === "recognised").map((item) => item.return_ref),
      world_mutated: returnModels.filter((item) => item.to_agency_ref === worldBinding.agency_ref && item.world_mutation_state === "applied").map((item) => item.return_ref),
    },
    constraints: worldBinding.constraints ?? {},
  };
}

export const EPI_LOGOS_ROOT_AGENCY_READING = Object.freeze({
  standing: ["Bimba", "Mono", "Possibility", "Determination", "0/1"],
  articulated: ["Pratibimba", "Poly", "Actualisation", "Labour", "1/0"],
  return: ["Return", "reconstitution", "1/1"],
});

File: crates/actuation-core/src/agency.rs
use crate::{
    ActionRef, ActivityRef, ActuationRef, AgencyIdentity, AgencyRef, AgentRef, AgentSessionRef,
    AuthorityRef, BoundsRef, DeterminationRef, Error, Extensions, ExternalRef, GrantRef, Invariant,
    InvocationRef, JourneyRef, NonEmpty, PlanRef, Record, RequestRef, Result, ReturnRef,
    ReturnRelationRef, RunRef, ScopeRef, Slot, WorldBindingRef, WorldRef,
};
use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub enum AgencySchema {
    #[default]
    #[serde(rename = "actuation.agency/v1")]
    V1,
}

macro_rules! invariant {
    ($ty:ty, [$($field:ident),*], $check:expr) => {
        impl Invariant for $ty {
            const FIELDS: &'static [&'static str] = &[$(stringify!($field)),*];
            fn extensions(&self) -> &Extensions { &self.extensions }
            fn check(&self) -> Result<()> { ($check)(self) }
        }
    };
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DeterminationKind {
    SelfDifferentiation,
    Delegation,
    Derivation,
    Federation,
}

#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum MetagencyOperation {
    ActualiseAgency,
    ConfigureAgency,
    DetermineAgency,
    ReintegrateReturn,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ReturnMode {
    Required,
    Optional,
    AutonomousTermination,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RecognitionState {
    Pending,
    Recognised,
    Rejected,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum MutationState {
    NotApplied,
    Applied,
}

/// Reading of actual supplied Return standing, not a Recognition decision.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ReturnStanding {
    Offered,
    Received,
    Rejected,
    Recognised,
    Reentered,
}

/// Explicit action lists describe bounded autonomy. Absence is not permission.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ActionStanding {
    Allowed,
    Denied,
    Unspecified,
}

pub type WorldConstraints = Record<WorldConstraintsFields>;
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct WorldConstraintsFields {
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub human_authored_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub security_policy_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub evidence_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub external_reality_refs: Slot<Vec<ExternalRef>>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
invariant!(
    WorldConstraintsFields,
    [
        human_authored_refs,
        security_policy_refs,
        evidence_refs,
        external_reality_refs
    ],
    |_: &Self| Ok(())
);

pub type WorldBinding = Record<WorldBindingFields>;
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct WorldBindingFields {
    pub schema: AgencySchema,
    pub binding_ref: WorldBindingRef,
    pub agent_ref: AgentRef,
    pub agency_ref: AgencyRef,
    pub world_ref: WorldRef,
    pub scope_ref: ScopeRef,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub purpose_ref: Slot<ExternalRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub continuity_ref: Slot<ExternalRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub determining_agency_ref: Slot<AgencyRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub return_relation_ref: Slot<ReturnRelationRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub bounds_refs: Slot<Vec<BoundsRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub authority_refs: Slot<Vec<AuthorityRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub constraints: Slot<WorldConstraints>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
invariant!(
    WorldBindingFields,
    [
        schema,
        binding_ref,
        agent_ref,
        agency_ref,
        world_ref,
        scope_ref,
        purpose_ref,
        continuity_ref,
        determining_agency_ref,
        return_relation_ref,
        bounds_refs,
        authority_refs,
        constraints
    ],
    |_: &Self| Ok(())
);

pub type RootScope = Record<RootScopeFields>;
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct RootScopeFields {
    pub schema: AgencySchema,
    pub scope_ref: ScopeRef,
    pub enclosing_world_ref: WorldRef,
    #[serde(flatten)]
    pub extensions: Extensions,
}
invariant!(
    RootScopeFields,
    [schema, scope_ref, enclosing_world_ref],
    |_: &Self| Ok(())
);

pub type MetagencyGrant = Record<MetagencyGrantFields>;
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct MetagencyGrantFields {
    pub schema: AgencySchema,
    pub grant_ref: GrantRef,
    pub agency_ref: AgencyRef,
    pub world_binding_ref: WorldBindingRef,
    pub authority_ref: AuthorityRef,
    pub operations: NonEmpty<MetagencyOperation>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub bounds_refs: Slot<Vec<BoundsRef>>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
invariant!(
    MetagencyGrantFields,
    [
        schema,
        grant_ref,
        agency_ref,
        world_binding_ref,
        authority_ref,
        operations,
        bounds_refs
    ],
    |_: &Self| Ok(())
);

pub type DelegatedAutonomy = Record<DelegatedAutonomyFields>;
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct DelegatedAutonomyFields {
    pub may_determine_within_bounds: bool,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub allowed_action_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub denied_action_refs: Slot<Vec<ExternalRef>>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
invariant!(
    DelegatedAutonomyFields,
    [
        may_determine_within_bounds,
        allowed_action_refs,
        denied_action_refs
    ],
    |_: &Self| Ok(())
);

pub type ReturnPolicy = Record<ReturnPolicyFields>;
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ReturnPolicyFields {
    pub mode: ReturnMode,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub return_relation_ref: Slot<ReturnRelationRef>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
invariant!(
    ReturnPolicyFields,
    [mode, return_relation_ref],
    |value: &Self| {
        if value.mode != ReturnMode::AutonomousTermination
            && value.return_relation_ref.value().is_none()
        {
            return Err(Error::new(
                "a required or optional Return must name its Return relation",
            ));
        }
        Ok(())
    }
);

pub type Determination = Record<DeterminationFields>;
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct DeterminationFields {
    pub schema: AgencySchema,
    pub determination_ref: DeterminationRef,
    pub kind: DeterminationKind,
    pub determining_agency_ref: AgencyRef,
    pub differentiated_agency_ref: AgencyRef,
    pub world_binding_ref: WorldBindingRef,
    pub bounds_refs: NonEmpty<BoundsRef>,
    pub delegated_autonomy: DelegatedAutonomy,
    pub return_policy: ReturnPolicy,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub parent_determination_ref: Slot<DeterminationRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub authority_refs: Slot<Vec<AuthorityRef>>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
invariant!(
    DeterminationFields,
    [
        schema,
        determination_ref,
        kind,
        determining_agency_ref,
        differentiated_agency_ref,
        world_binding_ref,
        bounds_refs,
        delegated_autonomy,
        return_policy,
        parent_determination_ref,
        authority_refs
    ],
    |value: &Self| {
        if value.kind == DeterminationKind::Federation
            && value
                .authority_refs
                .value()
                .is_some_and(|refs| !refs.is_empty())
        {
            return Err(Error::new(
                "federation does not grant determining authority; use explicit delegation",
            ));
        }
        Ok(())
    }
);

pub type ReturnProvenance = Record<ReturnProvenanceFields>;
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ReturnProvenanceFields {
    pub agency_lineage_refs: NonEmpty<AgencyRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub agent_refs: Slot<Vec<AgentRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub world_binding_refs: Slot<Vec<WorldBindingRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub authority_decision_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub authority_refs: Slot<Vec<AuthorityRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub bounds_refs: Slot<Vec<BoundsRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub activity_refs: Slot<Vec<ActivityRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub actuation_refs: Slot<Vec<ActuationRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub result_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub request_refs: Slot<Vec<RequestRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub agent_session_refs: Slot<Vec<AgentSessionRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub action_refs: Slot<Vec<ActionRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub invocation_refs: Slot<Vec<InvocationRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub plan_refs: Slot<Vec<PlanRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub journey_refs: Slot<Vec<JourneyRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub run_refs: Slot<Vec<RunRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub provider_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub harness_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub material_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub external_source_refs: Slot<Vec<ExternalRef>>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
invariant!(
    ReturnProvenanceFields,
    [
        agency_lineage_refs,
        agent_refs,
        world_binding_refs,
        authority_decision_refs,
        authority_refs,
        bounds_refs,
        activity_refs,
        actuation_refs,
        result_refs,
        request_refs,
        agent_session_refs,
        action_refs,
        invocation_refs,
        plan_refs,
        journey_refs,
        run_refs,
        provider_refs,
        harness_refs,
        material_refs,
        external_source_refs
    ],
    |_: &Self| Ok(())
);

pub type Return = Record<ReturnFields>;
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ReturnFields {
    pub schema: AgencySchema,
    pub return_ref: ReturnRef,
    pub determination_ref: DeterminationRef,
    pub from_agency_ref: AgencyRef,
    pub to_agency_ref: AgencyRef,
    pub difference_refs: NonEmpty<ExternalRef>,
    pub provenance: ReturnProvenance,
    pub received: bool,
    pub recognition_state: RecognitionState,
    pub world_mutation_state: MutationState,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub artifact_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub claim_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub evidence_refs: Slot<Vec<ExternalRef>>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
invariant!(
    ReturnFields,
    [
        schema,
        return_ref,
        determination_ref,
        from_agency_ref,
        to_agency_ref,
        difference_refs,
        provenance,
        received,
        recognition_state,
        world_mutation_state,
        artifact_refs,
        claim_refs,
        evidence_refs
    ],
    |value: &Self| {
        if !value.received && value.recognition_state != RecognitionState::Pending {
            return Err(Error::new(
                "a Return cannot be recognised or rejected before reception",
            ));
        }
        if value.world_mutation_state == MutationState::Applied
            && value.recognition_state != RecognitionState::Recognised
        {
            return Err(Error::new(
                "world mutation requires an explicitly recognised Return",
            ));
        }
        Ok(())
    }
);

impl WorldBindingFields {
    pub fn new(
        binding_ref: WorldBindingRef,
        agent_ref: AgentRef,
        agency_ref: AgencyRef,
        world_ref: WorldRef,
        scope_ref: ScopeRef,
    ) -> Self {
        Self {
            schema: AgencySchema::V1,
            binding_ref,
            agent_ref,
            agency_ref,
            world_ref,
            scope_ref,
            purpose_ref: Slot::Absent,
            continuity_ref: Slot::Absent,
            determining_agency_ref: Slot::Absent,
            return_relation_ref: Slot::Absent,
            bounds_refs: Slot::Absent,
            authority_refs: Slot::Absent,
            constraints: Slot::Absent,
            extensions: Extensions::new(),
        }
    }
}
impl WorldBinding {
    pub fn identity(&self) -> AgencyIdentity {
        AgencyIdentity {
            agent: self.fields().agent_ref.clone(),
            agency: self.fields().agency_ref.clone(),
        }
    }
    pub fn is_root_for(&self, scope: &RootScope) -> bool {
        self.fields().scope_ref == scope.fields().scope_ref
            && self.fields().world_ref == scope.fields().enclosing_world_ref
    }
    pub fn bounds(&self) -> &[BoundsRef] {
        self.fields()
            .bounds_refs
            .value()
            .map(Vec::as_slice)
            .unwrap_or(&[])
    }
    pub fn authorities(&self) -> &[AuthorityRef] {
        self.fields()
            .authority_refs
            .value()
            .map(Vec::as_slice)
            .unwrap_or(&[])
    }
}
impl RootScopeFields {
    pub fn new(scope_ref: ScopeRef, enclosing_world_ref: WorldRef) -> Self {
        Self {
            schema: AgencySchema::V1,
            scope_ref,
            enclosing_world_ref,
            extensions: Extensions::new(),
        }
    }
}
impl MetagencyGrant {
    /// This only reads the grant. Actualisation separately checks its exact
    /// WorldBinding, holder, authority source and bounds before admitting an act.
    pub fn includes(&self, operation: MetagencyOperation) -> bool {
        self.fields().operations.as_slice().contains(&operation)
    }
}
impl DelegatedAutonomy {
    pub fn action_standing(&self, action: &ExternalRef) -> ActionStanding {
        if self
            .fields()
            .denied_action_refs
            .value()
            .is_some_and(|refs| refs.contains(action))
        {
            ActionStanding::Denied
        } else if self
            .fields()
            .allowed_action_refs
            .value()
            .is_some_and(|refs| refs.contains(action))
        {
            ActionStanding::Allowed
        } else {
            ActionStanding::Unspecified
        }
    }
}
impl Return {
    pub fn standing(&self) -> ReturnStanding {
        let value = self.fields();
        if !value.received {
            ReturnStanding::Offered
        } else if value.world_mutation_state == MutationState::Applied {
            ReturnStanding::Reentered
        } else {
            match value.recognition_state {
                RecognitionState::Pending => ReturnStanding::Received,
                RecognitionState::Rejected => ReturnStanding::Rejected,
                RecognitionState::Recognised => ReturnStanding::Recognised,
            }
        }
    }
    /// Record reception only. No method here decides Recognition or applies a
    /// returned change to another owner's World.
    pub fn receive(self) -> Self {
        let mut fields = self.into_fields();
        fields.received = true;
        // Reception cannot invalidate an already admitted Return.
        Self::new(fields).expect("reception preserves the Return invariant")
    }
}

/// Validated aggregate lineage reading. As in the public v1 contract, this is
/// not a claim that its list is a unique, complete, ordered actualisation path.
/// R3's actualisation admission must require that stronger condition separately.
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(transparent)]
pub struct DeterminationLineage(NonEmpty<Determination>);
impl DeterminationLineage {
    pub fn new(items: NonEmpty<Determination>) -> Result<Self> {
        let by_ref: std::collections::BTreeMap<_, _> = items
            .as_slice()
            .iter()
            .map(|item| (&item.fields().determination_ref, item))
            .collect();
        for item in items.as_slice() {
            if let Some(parent_ref) = item.fields().parent_determination_ref.value() {
                let parent = by_ref.get(parent_ref).ok_or_else(|| {
                    Error::new(format!("missing parent determination {parent_ref}"))
                })?;
                if parent.fields().differentiated_agency_ref != item.fields().determining_agency_ref
                {
                    return Err(Error::new("recursive lineage must continue through its parent's differentiated Agency"));
                }
                if !parent
                    .fields()
                    .delegated_autonomy
                    .fields()
                    .may_determine_within_bounds
                {
                    return Err(Error::new(
                        "downward determination requires explicit delegated autonomy",
                    ));
                }
            }
        }
        Ok(Self(items))
    }
    pub fn as_slice(&self) -> &[Determination] {
        self.0.as_slice()
    }
}
impl<'de> Deserialize<'de> for DeterminationLineage {
    fn deserialize<D: serde::Deserializer<'de>>(
        deserializer: D,
    ) -> std::result::Result<Self, D::Error> {
        Self::new(NonEmpty::deserialize(deserializer)?).map_err(serde::de::Error::custom)
    }
}
